File: Data Structures & Algorithms/decode-ways/submission-2.cpp
class Solution {
public:
    int numDecodings(string s) {
        vector<int> dp(s.size()+1);
        int n = s.size();
        dp[n] = 1;

        for(int i = n-1;i>=0;i--){
            if(s[i]=='0') {
                dp[i]=0;
            } else {
                dp[i] = dp[i+1];
                if(i<n-1&& s[i] == '1' || s[i]=='2' && s[i+1]<'7'){
                    dp[i] += dp[i+2];
                }
            }
        }
        return dp[0];
        
    }
};

File: Data Structures & Algorithms/counting-bits/submission-2.py
class Solution:
    def countBits(self, n: int) -> List[int]:
        result = []
        
        for i in range(n + 1):
            # Count 1s in i using hamming weight trick
            num = i
            count = 0
            while num > 0:
                num = num & (num - 1)
                count += 1
            result.append(count)
        
        return result